### ISSUE TITLE
@wdio/mocha-framework@6.10.6 introduces TypeScript errors

### ISSUE THREAD

**Issue by redonkulus** [webdriverio/webdriverio#6247]
**Environment (please complete the following information):**
 - **WebdriverIO version:** 6.10.11
 - **Mode:** standalone
 - **If WDIO Testrunner, running sync/async:** async
 - **Node.js version:** v12.16.2
 - **NPM version:** 6.14.8
 - **Additional wdio packages used (if applicable):** 

**Describe the bug**
`@wdio/mocha-framework@6.10.6` causes the following TypeScript errors via `@types/jest`. I tried downgrading `@types/jest`, but the same errors occurred. They only went away when downgrading the `@wdio/mocha-framework` to version 6.10.4.

Seems like after this version was when @types/mocha was [added](https://github.com/webdriverio/webdriverio/commits/master/packages/wdio-mocha-framework/package.json).

I don't have a simple reproducible way to showcase the error right now. I just have to downgrade to the version that worked before. 

<details>
    <summary>error messages</summary>
<pre>
$ ./node_modules/.bin/tsc -p etc
node_modules/@types/mocha/index.d.ts:2581:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'beforeEach' must be of type 'Lifecycle', but here has type 'HookFunction'.

2581 declare var beforeEach: Mocha.HookFunction;
                 ~~~~~~~~~~

  node_modules/@types/jest/index.d.ts:34:13
    34 declare var beforeEach: jest.Lifecycle;
                   ~~~~~~~~~~
    'beforeEach' was also declared here.

node_modules/@types/mocha/index.d.ts:2599:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'afterEach' must be of type 'Lifecycle', but here has type 'HookFunction'.

2599 declare var afterEach: Mocha.HookFunction;
                 ~~~~~~~~~

  node_modules/@types/jest/index.d.ts:36:13
    36 declare var afterEach: jest.Lifecycle;
                   ~~~~~~~~~
    'afterEach' was also declared here.

node_modules/@types/mocha/index.d.ts:2615:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'describe' must be of type 'Describe', but here has type 'SuiteFunction'.

2615 declare var describe: Mocha.SuiteFunction;
                 ~~~~~~~~

  node_modules/@types/jest/index.d.ts:37:13
    37 declare var describe: jest.Describe;
                   ~~~~~~~~
    'describe' was also declared here.

node_modules/@types/mocha/index.d.ts:2636:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'xdescribe' must be of type 'Describe', but here has type 'PendingSuiteFunction'.

2636 declare var xdescribe: Mocha.PendingSuiteFunction;
                 ~~~~~~~~~

  node_modules/@types/jest/index.d.ts:39:13
    39 declare var xdescribe: jest.Describe;
                   ~~~~~~~~~
    'xdescribe' was also declared here.

node_modules/@types/mocha/index.d.ts:2650:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'it' must be of type 'It', but here has type 'TestFunction'.

2650 declare var it: Mocha.TestFunction;
                 ~~

  node_modules/@types/jest/index.d.ts:40:13
    40 declare var it: jest.It;
                   ~~
    'it' was also declared here.

node_modules/@types/mocha/index.d.ts:2664:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'test' must be of type 'It', but here has type 'TestFunction'.

2664 declare var test: Mocha.TestFunction;
                 ~~~~

  node_modules/@types/jest/index.d.ts:43:13
    43 declare var test: jest.It;
                   ~~~~
    'test' was also declared here.

node_modules/@types/mocha/index.d.ts:2671:13 - error TS2403: Subsequent variable declarations must have the same type.  Variable 'xit' must be of type 'It', but here has type 'PendingTestFunction'.

2671 declare var xit: Mocha.PendingTestFunction;
                 ~~~

  node_modules/@types/jest/index.d.ts:42:13
    42 declare var xit: jest.It;
                   ~~~
    'xit' was also declared here.
</pre>
</details>

**Comment by christian-bromann** [comment]
@redonkulus are you using Jest as your unit test framework?

**Comment by redonkulus** [comment]
Yes that is correct

**Comment by mohanraj-r** [comment]
I also ran into this error and other Type errors when upgrading "@wdio/local-runner", "@wdio/sync"

1. Error when upgrading "@wdio/local-runner" from "6.10.5" to "6.12.1"
```
ChromeDriver was started successfully.
[0-0] 2021-01-21T03:07:41.291Z ERROR @wdio/local-runner: Failed launching test session: TypeError: hooks.map is not a function
    at Object.executeHooksWithArgs (node_modules/@wdio/sync/build/executeHooksWithArgs.js:16:19)
    at Runner.run (node_modules/@wdio/runner/build/index.js:44:23)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
[0-0] FAILED
```

2. Error when upgrading  "@wdio/sync" from "6.10.4" to "6.11.0"
```
[0-0] Error in "AfterTest Hook"
hook.apply is not a function
[0-0] 2021-01-21T03:14:52.085Z ERROR @wdio/mocha-framework: Error in afterSuite hook: or: hooks.map is not a function
    at Object.executeHooksWithArgs (node_modules/@wdio/sync/build/executeHooksWithArgs.js:16:31)
    at Context.<anonymous> (node_modules/@wdio/mocha-framework/build/index.js:117:30)
    at callFn (node_modules/mocha/lib/runnable.js:364:21)
    at Hook.Runnable.run (node_modules/mocha/lib/runnable.js:352:5)
    at next (node_modules/mocha/lib/runner.js:510:10)
    at Immediate.<anonymous> (node_modules/mocha/lib/runner.js:572:5)
    at processImmediate (internal/timers.js:461:21)
[0-0] 2021-01-21T03:14:52.086Z ERROR @wdio/sync: TypeError: hook.apply is not a function
    at execHook (node_modules/@wdio/sync/build/executeHooksWithArgs.js:21:31)
    at Runner.run (node_modules/@wdio/runner/build/index.js:89:24)
[0-0] 2021-01-21T03:14:52.086Z ERROR @wdio/sync: TypeError: hook.apply is not a function
    at execHook (node_modules/@wdio/sync/build/executeHooksWithArgs.js:21:31)
    at Runner.run (node_modules/@wdio/runner/build/index.js:89:24)
[0-0] 2021-01-21T03:14:52.086Z ERROR @wdio/sync: TypeError: hook.apply is not a function
    at execHook (node_modules/@wdio/sync/build/executeHooksWithArgs.js:21:31)
    at Runner.run (node_modules/@wdio/runner/build/index.js:89:24)
```

Looks like the TS2304 is not an isolated incident
* [error TS2304: Cannot find name 'Element' · Issue #5719 · webdriverio/webdriverio](https://github.com/webdriverio/webdriverio/issues/5719)
* [mocha type definitions conflict with @types/jest · Issue #6690 · cypress-io/cypress](https://github.com/cypress-io/cypress/issues/6690)

I tried some suggestions from [Handling type declarations clash in Typescript](https://medium.com/@elenasufieva/handling-type-declarations-clash-in-typescript-b05b10723f47). But both `--skipLibCheck` and `typeRoots` overrides resulted in different errors.

But I am able to upgrade some other wdio packages to the latest versions without errors
```
        "@wdio/cli": "6.12.1",
        "@wdio/spec-reporter": "6.11.0",
        "webdriverio": "6.12.1"
```

**Comment by christian-bromann** [comment]
@mohanraj-r can you provide a minimal reproducible example?

**Comment by mohanraj-r** [comment]
@christian-bromann I am sorry for the delay - I have been meaning to get to this. 
In the meanwhile you can reproduce the errors by cloning the repo https://github.com/salesforce/sa11y and running `yarn && yarn install:update` and choosing the wdio related pkgs to update.

**Comment by redonkulus** [comment]
Just tried upgrading to v7 of all the webdriver packages and got the same error again. @christian-bromann were you able to look at @mohanraj-r repo?

**Comment by mohanraj-r** [comment]
Same here, tried upgrading to v7 all the webdriverIO packages together and also one by one - not able to upgrade because of the errors.

**Comment by redonkulus** [comment]
I ended up using the ` "skipLibCheck": true,` option in my `tsconfig.json` to work around the issue for now.

**Comment by christian-bromann** [comment]
@redonkulus @mohanraj-r I am not sure how I can best help you. The errors you mention in your [last comment](https://github.com/webdriverio/webdriverio/issues/6247#issuecomment-764211204) aren't TypeScript errors, these are actual JavaScript issues that should never happen if WebdriverIO is configured correctly. Maybe we are talking about multiple problems at once. To make helping you easier I suggest:

- to try to upgrade to v7 where we overhauled the TypeScript setup
- create a minimal reproducible example to isolate any of the problems

What do you all think?

**Comment by mohanraj-r** [comment]
Thanks @christian-bromann 

I did try to upgrade to v7 again recently as of v7.3.0 and now again to 7.5.1
* I refactored code to account for renamed types etc
* I created a tsconfig specific to wdio packages in the mono-repo
```json
{
  "extends": "./tsconfig.common.json",
  "compilerOptions": {
      "skipLibCheck": true,
      "types": ["node", "webdriverio", "webdriverio/async", "@wdio/mocha-framework", "expect-webdriverio"]
  }
}
```
I am following the advice from the doc:
> Please avoid importing webdriverio or @wdio/sync explicitly. WebdriverIO and WebDriver types are accessible from anywhere once added to types in tsconfig.json
* I added config as suggested in the docs
```

    autoCompileOpts: {
        autoCompile: true,
        // see https://github.com/TypeStrong/ts-node#cli-and-programmatic-options
        // for all available options
        tsNodeOpts: {
            transpileOnly: true,
            project: 'tsconfig.json',
        },
        // tsconfig-paths is only used if "tsConfigPathsOpts" are provided, if you
        // do please make sure "tsconfig-paths" is installed as dependency
        tsConfigPathsOpts: {
            baseUrl: './',
        },
    },
```

But even after the above steps I keep running into errors. The latest set of errors I am facing is `Cannot find name 'browser'`
```
$ wdio run wdio.conf.js --suite wdio --suite integration --suite browserLib < /dev/null

Execution of 3 spec files started at 2021-04-22T21:24:57.365Z

Starting ChromeDriver 90.0.4430.24 (4c6d850f087da467d926e8eddb76550aed655991-refs/branch-heads/4430@{#429}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
[0-1] RUNNING in chrome - /packages/test-integration/__tests__/wdio.test.ts
[0-2] RUNNING in chrome - /packages/browser-lib/__tests__/browser-lib.test.ts
[0-0] RUNNING in chrome - /packages/wdio/__tests__/wdio.test.ts
[0-0] 2021-04-22T21:25:02.962Z ERROR @wdio/runner: Error: Unable to load spec files quite
 likely because they rely on `browser` object that is not fully initialised.
`browser` object has only `capabilities` and some flags like `isMobile`.
Helper files that use other `browser` commands have to be moved to `before` hook.
Spec file(s): /packages/wdio/__tests__/wdio.test.ts
Error: TSError: ⨯ Unable to compile TypeScript:
packages/wdio/__tests__/wdio.test.ts(32,111): error TS2322: Type 'number' is not assignable to type '{ label: number; sent: () => any; trys: any[]; ops: any[]; }'.
packages/wdio/__tests__/wdio.test.ts(62,46): error TS2304: Cannot find name 'browser'.
...
``` 

* Have pushed all my changes to https://github.com/mohanraj-r/sa11y/tree/upgrade_wdio_v7
  - if you want to give it a try : 'yarn build; yarn test:wdio'

* Am planning to try the following to see if it helps when I get some cycles next
  - use the wdio cli to regenerate the wdio config from scratch 
  - remove sync mode and if the simplification helps

**Comment by erwinheitzman** [comment]
One thing that stands out to me is the use of the following types `"webdriverio", "webdriverio/sync","webdriverio/async"`.
You only need `webdriverio/sync` or `webdriverio/async` out of these three. Dropping the other two might resolve the issue of conflict.

**Comment by mohanraj-r** [comment]
Thanks @erwinheitzman 
FYI I am building on top of WDIO to offer both sync and async APIs - that is why I thought I needed both sync and async. 

Anyways, I tried with just "webdriverio/sync" and then also with just "webdriverio/async" instead - same error in both cases `Error:  Unable to load spec files quite likely because they rely on `browser` object that is not fully initialised.`

**Comment by erwinheitzman** [comment]
Well the real issue is probably the `Error: TSError: ⨯ Unable to compile TypeScript` so first fix this ;)

**Comment by mohanraj-r** [comment]
Thanks, Yes I noticed that - but the issues don't come when I run `tsc` directly on the project and they certainly didn't come with WDIO v6 - they are only surfacing when upgrading to v7. 

When I run `tsc` specifically on the wdio test file, a different set of errors surface pointing to conflicts with existing packages - which could well be the source of the issue. Not sure why these 
```bash
$ yarn tsc packages/wdio/__tests__/wdio.test.ts
yarn run v1.22.10

node_modules/@types/jasmine/index.d.ts:26:1 - error TS6200: Definitions of the following
identifiers conflict with those in another file: beforeAll, beforeEach, afterAll, afterEach, describe, fdescribe, xdescribe, it, fit, xit, expect, DEFAULT_TIMEOUT_INTERVAL, CustomMatcherFactory, CustomEqualityTester

26 type ImplementationCallback = jasmine.ImplementationCallback;
   ~~~~

  node_modules/@types/jest/index.d.ts:34:1
    34 declare var beforeAll: jest.Lifecycle;
       ~~~~~~~
    Conflicts are in this file.

node_modules/@types/jasmine/index.d.ts:33:18 - error TS2300: Duplicate identifier 'describe'.

33 declare function describe(description: string, specDefinitions: () => void): void;
                    ~~~~~~~~

  node_modules/@types/mocha/index.d.ts:2618:13
    2618 declare var describe: Mocha.SuiteFunction;
                     ~~~~~~~~
    'describe' was also declared here.

node_modules/@types/jasmine/index.d.ts:47:18 - error TS2300: Duplicate identifier 'xdescribe'.

47 declare function xdescribe(description: string, specDefinitions: () => void): void;
                    ~~~~~~~~~

  node_modules/@types/mocha/index.d.ts:2639:13
    2639 declare var xdescribe: Mocha.PendingSuiteFunction;
                     ~~~~~~~~~
    'xdescribe' was also declared here.
...
```

Again these errors doesn't occur with wdio v6
```
$ yarn tsc packages/wdio/__tests__/wdio.test.ts
yarn run v1.22.10

✨  Done in 6.89s.
```

`Cannot find name 'browser'` is a real head scratcher and is the most occurring error - isn't the wdio test runner supposed to create the global browser object - how come it is not found then?

**Comment by mkillgallon** [comment]
@redonkulus did you manage to solve the "error TS2304: Cannot find name 'browser'." ? I'm getting the same problem.

**Comment by redonkulus** [comment]
@mkillgallon I was able to upgrade to 7.x without issue, so maybe it fixed itself.

**Comment by christian-bromann** [comment]
@redonkulus @mkillgallon @mohanraj-r can someone provide a reproducible example using the latest WebdriverIO version?

**Comment by mohanraj-r** [comment]
I tried again today with latest version of wdio pkgs. I think the error comes from type collision from jest-jasmine and wdio-mocha defining some of the same types. So a monorepo project with both jest and wdio as common dev deps at the root level probably results in the errors. I tried moving the wdio deps from the monorepo root level to the individual pkg level - but I still need to have wdio cli, local runner etc at root level for the npm test scripts invoking the wdio bin to work. So couldn't get it to work as expected 😞 

I will try to extract and produce an example with both jest and wdio deps in a sample project when I can get some time next.
